src/ripple/core/impl/Stoppable.cpp

```diff
@@ -85,11 +85,11 @@ void Stoppable::onStart ()
 
 void Stoppable::onStop ()
 {
-    stopped();
```

Comment:
I'd expect `stopped` to be called from `onStop` and not from `onChildrenStopped`. If a class derives from Stoppable, then it overrides `onChildrenStopped` but not `onStopped` it has to know to called `stopped()`? That doesn't seem intuitive.

I think the motivation for this change is to simplify the implementation of the `JobQueue`, is that right? I think I'd prefer to keep `Stoppable::onStop` as it is, and maybe add a comment in `onStop` to say why the base class does the wrong thing for this case.

If we do decide to keep this change, we need to update the docs on `onChildrenStopped` and `onStopped` to reflect the new behavior (the current docs says `onStopped` calls `stopped` and `onChildrenStopped` does nothing)

Reply:
> If a class derives from `Stoppable`, and it overrides `onChildrenStopped` but not `onStop`, then it has to know to called `stopped()`? That doesn't seem intuitive.

I think this is just a matter of naming. We can change the names, and I'd be happy to. It's a harder sell to change public API method names, so that's why I left them alone. I think better names would be `beforeChildrenStopped` and `afterChildrenStopped`. The documentation suggests that a parent can expect to no longer be called by its children after they have stopped, which suggests to me that that is the only safe time for a parent to stop, which means that's when it should call `stopped()`.

> I think the motivation for this change is to simplify the implementation of the `JobQueue`, is that right?

The primary motivation is to simplify the uses of the API, which includes `JobQueue`. I think the default implementation is wrong, considering the semantics of the class:

- Children depend on their parents.
- Dependents should be stopped before dependencies are stopped.
- `onChildrenStopped` is called after all of a `Stoppable`'s dependents have stopped.
- Thus, `onChildrenStopped` is when `stopped()` should be called.

> If we do decide to keep this change, we need to update the docs

Indeed, that was my oversight. Thanks for reminding!

Reply:
I have to think about this one. Unlike the other changes in this PR, this changes in a critical API.

Reply:
Yeah, I suspect there's something thread-related wrong with this pull request.  I fired off a start-sync-for-60-seconds-stop loop on this branch.  The 48th start-stop died with a crash.  The actual crash was here:
```
11  rippled                       	0x000000010ac2911b ripple::OverlayImpl::collect_metrics() + 139 (vector:657)
12  rippled                       	0x000000010b06b31a beast::insight::detail::StatsDCollectorImp::on_timer(boost::system::error_code) + 234
```
but it is also interesting to note that the `ApplicationImp` destructor was running at the time:
```
Thread 0:: rippled: main  Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	0x00007fff63f41f06 __psynch_mutexwait + 10
1   libsystem_pthread.dylib       	0x00007fff63ffed52 _pthread_mutex_firstfit_lock_wait + 96
2   libsystem_pthread.dylib       	0x00007fff63ffc4cd _pthread_mutex_firstfit_lock_slow + 222
3   libc++.1.dylib                	0x00007fff6103c0bf std::__1::recursive_mutex::lock() + 9
4   rippled                       	0x000000010b05cd6b beast::insight::detail::StatsDHookImpl::~StatsDHookImpl() + 59 (List.h:151)
5   rippled                       	0x000000010ac2950e ripple::OverlayImpl::~OverlayImpl() + 190 (memory:3587)
6   rippled                       	0x000000010ac29c2e ripple::OverlayImpl::~OverlayImpl() + 14 (OverlayImpl.cpp:183)
7   rippled                       	0x000000010af66525 ripple::ApplicationImp::~ApplicationImp() + 325 (memory:2649)
8   rippled                       	0x000000010af5dafe ripple::ApplicationImp::~ApplicationImp() + 14 (Application.cpp:250)
```
If it would be useful I can share the scripts I use to automate that test loop.  I run them on a Mac, so you'd have to tweak them...

Reply:
@scottschurr can you try to the same test against the version on top of which this PR is built on.

Reply:
@nbougalis, yes I'll try the same against -rc3.   FWIW a few years back, when I was working on thread safety, I was able to get well over 1000 start-sync-stop loops with no crashes.  In my opinion if we're not able to attain that then there's some work to do.

Reply:
@nbougalis, I ran the same test against -rc3 and it showed approximately the same level of stability as this branch.  This time there was a crash on the 38th iteration.  Again the crash was in `OverlayImpl::collect_metrics()`:

```
9   libc++.1.dylib                	0x00007fff789b50e7 std::__1::__throw_system_error(int, char const*) + 77
10  libc++.1.dylib                	0x00007fff7896f96d std::__1::mutex::lock() + 29
11  rippled                       	0x00000001083366bb ripple::OverlayImpl::collect_metrics() + 139 (vector:657)
12  rippled                       	0x0000000108778ada beast::insight::detail::StatsDCollectorImp::on_timer(boost::system::error_code) + 234
13  rippled                       	0x0000000108778e53 boost::asio::detail::wait_handler<std::__1::__bind<void (beast::insight::detail::StatsDCollectorImp::*)(boost::system::error_code), beast::insight::detail::StatsDCollectorImp*, std::__1::placeholders::__ph<1> const&>, boost::asio::detail::io_object_executor<boost::asio::executor> >::do_complete(void*, boost::asio::detail::scheduler_operation*, boost::system::error_code const&, unsigned long) + 227 (type_traits:4302)
14  rippled                       	0x00000001074132c8 boost::asio::detail::scheduler::do_run_one(boost::asio::detail::conditionally_enabled_mutex::scoped_lock&, boost::asio::detail::scheduler_thread_info&, boost::system::error_code const&) + 920
```
It looks like the `ApplicationImp` destructor is running in another thread and it's trying to destroy `StatsDHookImpl`:
```
3   libc++.1.dylib                	0x00007fff789740bf std::__1::recursive_mutex::lock() + 9
4   rippled                       	0x000000010876a52b beast::insight::detail::StatsDHookImpl::~StatsDHookImpl() + 59 (List.h:151)
5   rippled                       	0x0000000108336aae ripple::OverlayImpl::~OverlayImpl() + 190 (memory:3587)
6   rippled                       	0x00000001083371ce ripple::OverlayImpl::~OverlayImpl() + 14 (OverlayImpl.cpp:183)
7   rippled                       	0x0000000108673d45 ripple::ApplicationImp::~ApplicationImp() + 325 (memory:2649)
8   rippled                       	0x000000010866b31e ripple::ApplicationImp::~ApplicationImp() + 14 (Application.cpp:250)
9   rippled                       	0x00000001086696bd ripple::run(int, char**) + 16781 (memory:2649)
```
My take is that we shouldn't be making any changes to `Stoppable` until the stability is fixed.  Or maybe another way to put it is that any changes to `Stoppable` should be specifically associated with fixing this stability issue.  And the stability should be significantly improved by any changes to `Stoppable`.

The stability problem may not be in `Stoppable`, but until the stability is fixed we have no way of validating any threading-related changes.

Reply:
I'm digging deeper into `Stoppable` to analyze the start and stop order. I'll see if I can find the cause of this.

Reply:
So, I spent some time looking at this. We _recently_ changed this code (https://github.com/ripple/rippled/commit/b784988caf757d919abe6ddc30fe62ee28a9d779) and I suspect that the problem is that the collector keeps trying to collect metrics.

Tagging @mayurbhandary for visibility.